### PR TITLE
feat: flow cli for  distributed

### PR DESCRIPTION
**Pull request by discord9**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

Add command line options for flownode, but didn't add FrontendInvoker for it now since PR size is alreay too large

Please explain IN DETAIL what the changes are in this PR and why they are needed:

## Summary

- ~~***Breaking change***: make all cli option no longer accept `metasrv-addr` as option but only `metasrv-addrs`(docs change at https://github.com/GreptimeTeam/docs/pull/1030)~~ see #4239 
- refactor `FlownodeBuilder`
- add cli options for `flownode`
- impl flownode build for `flownode` cli options

NOTE: flownode service is still not working, since the `FrontendInvoker` is not yet set, so the flow's computing result will not be send back to frontend yet, this is planned in a future PR

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
  - Introduced a new module for managing Greptime Flownode, including starting, stopping, and configuration.
  - Added a new `Flownode` subcommand to enhance command-line functionality.

- **Bug Fixes**
  - Improved error handling for server start, shutdown, and address parsing.
  
- **Refactor**
  - Renamed `FlownodeManager` to `FlowWorkerManager` and updated related references.
  - Streamlined import paths and reorganized module structures for better maintainability.

### CODE REVIEW

File: src/cmd/src/error.rs [hunk 1]
```diff
@@ -87,6 +87,20 @@ pub enum Error {
         source: datanode::error::Error,
     },
 
+    #[snafu(display("Failed to startflownode"))]
```
Comment (fengjiachun) [inline src/cmd/src/error.rs:90]:
```suggestion
    #[snafu(display("Failed to start flownode"))]
```

File: src/flow/src/heartbeat.rs [hunk 2]
```diff
@@ -196,3 +222,38 @@ impl HeartbeatTask {
         }
     }
 }
+
+/// Create metasrv client instance and spawn heartbeat loop.
+pub async fn new_metasrv_client(
+    node_id: u64,
+    meta_config: &MetaClientOptions,
+) -> Result<MetaClient, Error> {
+    let cluster_id = 0; // TODO(hl): read from config
```
Comment (fengjiachun) [inline src/flow/src/heartbeat.rs:231]:
```suggestion
    let cluster_id = 0; // TODO(discord9): read from config
```
:)

**Review by fengjiachun — APPROVED** [review]

**Review by killme2008 — COMMENTED** [review]
Good job! After this PR, can we add some integration tests for flow into `tests-integration`?

File: src/cmd/src/flownode.rs [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+
+#[derive(Debug, Parser, Default)]
+struct StartCommand {
+    #[clap(long)]
+    node_id: Option<u64>,
+    #[clap(long)]
+    rpc_addr: Option<String>,
+    #[clap(long)]
+    rpc_hostname: Option<String>,
+    #[clap(long, aliases = ["metasrv-addr"], value_delimiter = ',', num_args = 1..)]
```
[the remark this answers is not in the document]
Comment (discord9) [reply src/cmd/src/flownode.rs:128]: > Yep, it's breaking change. But I think it's better to break earlier.

I thought the aliases here are just for compatibility, making both `metasrv-addr` and `metasrv-addrs` acceptable, I don't know if it is necessary to remove this alias and only accept `metasrv-addrs` as a option?

[the remark this answers is not in the document]
Comment (killme2008) [reply src/cmd/src/flownode.rs:128]: I got it. I don't the new service still needs this compatibility, just remove it please.

**Comment by discord9** [comment]
> Good job! After this PR, can we add some integration tests for flow into `tests-integration`?

Will add some tests&docs update after impl `FrontendInvoker` and make flownode actually runnable in next PR

File: src/flow/src/heartbeat.rs [hunk 4]
```diff
@@ -82,6 +96,19 @@ impl HeartbeatTask {
         Ok(())
     }
 
+    pub fn shutdown(&self) -> Result<(), Error> {
```
Comment (WenyXu) [inline src/flow/src/heartbeat.rs:99]: nit:
```suggestion
    pub fn shutdown(&self) {
```

File: src/cmd/src/flownode.rs [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+            .context(BuildCacheRegistrySnafu)?
+            .build(),
+        );
+
+        let catalog_manager = KvBackendCatalogManager::new(
+            opts.mode,
+            Some(meta_client.clone()),
+            cached_meta_backend.clone(),
+            layered_cache_registry.clone(),
+        );
```
[the remark this answers is not in the document]
Comment (discord9) [reply src/cmd/src/flownode.rs:267]: invalidation is added, although I'm not sure if I impl that correctly@WenyXu @fengjiachun  please help check that?: https://github.com/GreptimeTeam/greptimedb/pull/4226/commits/6cd9c7592f8c0d9957bc418fba235fc02e3c7347

**Review by WenyXu — APPROVED** [review]
LGTM

File: src/cmd/src/flownode.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (discord9) [reply src/cmd/src/flownode.rs:128]: There is some complication after change that, considering this PR is already too large, will open another PR for change this option

**Pull request merged**